banzai/bpm.py

```diff
@@ -8,7 +20,72 @@ def group_by_keywords(self):
         return None
 
     def do_stage(self, images):
+        images_to_remove = []
         for image in images:
-            bpm_slices = array_utils.array_indices_to_slices(image.bpm)
-            image.bpm[image.data[bpm_slices] >= float(image.header['SATURATE'])] = 2
+            try:
+                self.add_bpm_to_image(image)
+            except MissingBPMError:
+                if self.pipeline_context.no_bpm:
+                    logger.info('BPM misisng but not required, falling back to empty BPM', image=image)
+                    pass
+                else:
+                    logger.error("Can't add BPM to image, stopping reduction", image=image)
+                    images_to_remove.append(image)
+                    continue
+            flag_bad_pixels(image)
+        for image in images_to_remove:
+            images.remove(image)
         return images
+
+    def add_bpm_to_image(self, image):
+        # Get the BPM filename
+        bpm_filename = self.get_bpm_filename(image)
+        # Check if file is missing
+        if bpm_filename is None:
+            logger.warning('Unable to find BPM in database', image=image)
+            raise MissingBPMError
```

Comment:
See above comment. I think this should load a dummy bpm, not raise an exception.

Reply:
The exception is now removed